### ISSUE TITLE
Completion Flag Already Set on Torrent Only 95% Downloaded

### ISSUE THREAD

**Issue by CrafterLaughter** [BiglySoftware/BiglyBT#899]
Windows 10 x64
BiglyBT Ver 1.8.0

Not a reproducible bug, yet I don't see a user forum for BiglyBT, so I'm hoping someone can be kind to offer some guidance.  I serve a rural community in Africa with intermittent internet and roughly 1GB of bandwidth each day.  I host a Vuze torrent tracker for them by which they've downloaded a dozen 10G torrents from my tracker without issue.  They downloaded an 11GB torrent from my tracker the past 2 weeks and something unusual happened.  The completion flag is set on the torrent despite it being only 95% downloaded.  The console has the following entries.

> {disk} valid=true,not done=270,done=4969,started=0,recheck=0,rc all=false,full=false; 	| TorrentDM: 'world_possible_linux_server.7z'
> 
> {core} Completion flag already set on "world_possible_server.7z", skip move-on-completion behaviour.; 	| TorrentDLM: 'world_possible_linux_server.7z'

The first thing I did was ensure the checksum of the 11GB file in question is unchanged on the tracker PC.  Then I downloaded the torrent from the tracker with another PC.  It succeeded and the checksum matched so I don't believe it's any corruption on the tracker PC.  It remains in seeding status despite having force re-checked.  The console indicated dozens of pieces that fail hash check which doesn't surpise me given their internet connection gets interrupted dozens of times over the course of a download like this.  I might guess those pieces automatically get discarded and downloaded anew?

Can anyone offer suggestions as to removing the completion flag from a torrent or anything else to try and reconnaissance a torrent in this state.  I skimmed the debug logs and nothing stood out to me and I expect I just don't know what to look for.  I can upload any logs necessary to figure this out.  I immensely appreciate any help as it would allow us to avoid delaying a school computer lab deployment a couple more weeks from having to start over on the download.

**Comment by parg** [comment]
So to recap you have a user stuck with a 95% complete download that fails to make any further downloading progress, even though you know there is a seed running for that download. Also you have attempted a 'force recheck' on both seed and this downloader and that hasn't moved things?

The 'completion flag' itself won't prevent download, it is used to prevent 'move on complete' actions being attempted twice (for example). The fact that it is set seems to imply that the download at one point completed but then subsequently reverted to an incomplete state - perhaps the recheck while seeding threw up some bad pieces. If you get a lot of these it might imply a disk issue as at one point they were verified as correct.

Do you see any peer connections from the 95% complete client to the seed? It may be that the downloader has blocked the seed due to repeat hash fails.  Right-click on the status bar and select 'IPFilter Status' from the 'Show' menu (if not already visible). Then you can double-click on the status info that is shown ('IPs n: n/n/n') to see if any peers have been blocked.

**Comment by CrafterLaughter** [comment]
I'd only run a force re-check on the client.  Thank you for reminding me I needed to do it on the tracker as well.  Unfortunately doing so didn't change anything.  I ran a diagnostic to check all the sectors on the hard drive and that came back clean for now.  The client being clocked due to repeat hash fails makes total sense.  However, I actually disabled IP filtering when trying to fix this.

I did wind up discovering something totally unexpected. For some reason on the client, the priority got switched to "do not download" and as soon as I toggled that back to "normal" priority the torrent status changed from "seeding" back to "downloading."  By chance are there reason(s) an inbuilt client could automatically switch a torrent priority to "do not download"?  Sadly, despite the torrent switching back to downloading status it actually hasn't downloaded anything and the console just scrolls through re-scanning the pieces in a loop.

Attached are a few screenshots.  Every few minutes I see the client and tracker showing each other on the peer connections tab as pictured in the screenshots.

![image](https://user-images.githubusercontent.com/23381273/53688771-33394480-3d06-11e9-8d08-844cec8fd351.png)

![image](https://user-images.githubusercontent.com/23381273/53688775-3fbd9d00-3d06-11e9-8347-66ad01c530de.png)

![image](https://user-images.githubusercontent.com/23381273/53688776-44825100-3d06-11e9-9171-c7e6d14a78fb.png)

![image](https://user-images.githubusercontent.com/23381273/53688779-4a783200-3d06-11e9-9ba4-0b6672cd5fe9.png)

![image](https://user-images.githubusercontent.com/23381273/53688783-611e8900-3d06-11e9-9f21-909b0af2d25a.png)

![image](https://user-images.githubusercontent.com/23381273/53688784-64197980-3d06-11e9-860e-78c6c274e0e8.png)

**Comment by parg** [comment]
Hmm, did it switch back to 'do not download' again or is that screenshot from before you switched it back? 
Once a download is added to BiglyBT there isn't anything that would automatically switch a file to a DND state that I can think of - there is the checkbox in the FilesView that could be unselected by accident I suppose.

The banning of peers for sending bad data is separate from the explicit IPFilter itself so it is worth double checking that the seed hasn't been added to the 'banned list'

Do you ever see successful connections between the seed and the other client? If you stop and start the download in the incomplete client the log should show then initial tracker response and subsequent attempted peer connection to the seed (hopefully!) - there might be something useful in that trace.

**Comment by CrafterLaughter** [comment]
That screenshot was prior to my switching it back to normal priority.  Users do inexplicable things sometimes, so I'll assume my guy inadvertently deselected it in Files view.

I'm glad to know the banning of peers for sending bad data is separate from the explicit IPFilter itself.  I just kept messing with the IP Filters options and didn't realize that double left clicking on the status area for IPFilter shows the actual ban list.  Though I didn't see any IPs in the banned list, I may have inadvertently cleared my tracker from the ban list by my earlier meddling in IP Filters options and deselecting "save blocked IP details across restarts."  I suspect that's the case because that's the last thing I did before I noticed the transfer has resumed.

Parg...You're the man.  We immensely appreciate your helpful suggestions.  Have an awesome week!

![image](https://user-images.githubusercontent.com/23381273/53726934-49d9bb80-3e2c-11e9-9e26-324f2503ded9.png)

![image](https://user-images.githubusercontent.com/23381273/53726946-4e9e6f80-3e2c-11e9-8dbb-fc76436e508c.png)

![image](https://user-images.githubusercontent.com/23381273/53726919-3fb7bd00-3e2c-11e9-97b7-e7628a1764bb.png)
